### ISSUE TITLE
Travail sur les titres colan 

### ISSUE THREAD

**Issue by hls1138** [hls1138/IntroMachineLearning-#1]
### %matplotlib inline
import nltk
import nltk.corpus
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import os
with open(os.path.join("C://Users//henry//Desktop//TousLesTitresCL.txt")) as titres:
    titres = [line.strip() for line in titres.readlines()]
#Pour vérificication (structure, longueur et type)
#print(titres)
#len(titres)
#type(titres)
#je n'ai pas réussi à utiliser les stop words francais issus de nltk 
#pour l'étape suivante nous allons utiliser une liste de stopwords pour enlever de notre tfidfvectorizer les mots les plus courents                 
with open('C://Users//henry//Desktop//StopWordsFr.txt', 'r', encoding="utf-8") as StopWords:
   StopWords=[line.strip() for line in StopWords.readlines()]
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer( encoding='utf-8', max_df=1, stop_words=StopWords)
vectorizer.fit(titres)
#vectorizer.vocabulary_ mis en commentaire pour plus de lisibilité
#nous avons maintenant un dictionnaire de features
vv= vectorizer.vocabulary_
type(vv)
#Un tel type de visualisation est peu utile compte tenu du nombre de mots dans le dictionnaire. 
import matplotlib.pyplot as plt
plt.bar(range(len(vv)), list(vv.values()), align='center')
plt.xticks(range(len(vv)), list(vv.keys()))
plt.show()
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(lowercase=False, stop_words=StopWords)
tfidf_vectorizer.fit(titres)
#nous traitons le vocabulaire des titres
#tfidf_vectorizer.vocabulary_ 
#mis en commentaire pour plus de lisibilité
#Affichage des mots les plus fréquents (selon TFxIDF) :
X_titres = tfidf_vectorizer.transform(titres)
features_titres = tfidf_vectorizer.get_feature_names()

D_tf = X_titres.toarray()
tf_sum = np.sum(D_tf, axis=0)
top_tfidf_feats(tf_sum, features_titres)
def top_mean_feats(Xtr, features, grp_ids=None, min_tfidf=0.1, top_n=25):
  #  ''' Return the top n features that on average are most important amongst documents in rows
#     indentified by indices in grp_ids. '''
    if grp_ids:
        D = Xtr[grp_ids].toarray()
    else:
        D = Xtr.toarray()
    ''' D[D < min_tfidf] = 0 '''
    tfidf_means = np.mean(D, axis=0)
    return top_tfidf_feats(tfidf_means, features, top_n)

top_mean_feats(X_titres, features_titres)
#un document à represéenter comme un ensemble de mots clefs on perd l'ordre sac de mots.
#le VSM est un outil puissant car on peut apliquer des maths dessus 
#l'ajout de métadonnées (données sur le texte) pour le tagging ( au niveau des mots, des tokens des noms des sentiments)
#plus complexe on peut faire des représentations sous forme d'arbre syntaxique.
#Chaque représentation implique une manière différente de comparer les documents
#on cherche à obtenir un index inversée terme (ligne et doc commande) avce la fréquence en intersection ce tableau partagent le meme vocabulaire
#matrice avec bcp de 0 alors on à une matrice creuse. tf = term fréquency
#plusieurs étapes pour construire la matricre 
#1 on tokenize (espace et poncuation ) 
#2 on construit un vocabulaire over tout le documemt.+ lowercasing et etre un peu vigilant
#3 sparse (creuse) matrice encoding (avec les numéros par tf)
#td5 classification de données textuelles
lines_bag_of_words = vectorizer.transform(titres).toarray()
lines_bag_of_words.shape
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()

lr.fit(lines_bag_of_words, titres)
lr.predict(titres[:100])
from sklearn.naive_bayes import MultinomialNB
classifier = MultinomialNB().fit(lines_bag_of_words, titres)
classifier.predict(titres[:100])
from sklearn.cluster import KMeans
from sklearn.metrics import adjusted_rand_score
clustering = KMeans(n_clusters=8, init='k-means++', max_iter=1000, n_init=1)
clustering.fit(X_titres)
#nous souhaitons affichcher les clusters avec les termes les plus proches les uns des autres dans les titres
print("Termes les plus fréquents par clusters")
centroides = model.cluster_centers_.argsort()[:, ::-1]
termes = features_titres
for i in range(8):
    print("Cluster %d:" % i),
    for ind in centroides[i, :10]:
        print(' %s' % termes[ind]),
    print
